src/AzureAuth/Ado/AdoToken.cs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.AzureAuth.Ado
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    using Microsoft.Office.Lasso.Interfaces;
+
+    /// <summary>
+    /// Business logic for either getting an ADO PAT from <see cref="IEnv"/>
+    /// or doing normal AAD Authentication.
+    /// </summary>
+    public static class AdoToken
+    {
+        private static readonly IEnumerable<string> PatEnvVars = new[]
+        {
+            EnvVars.AdoPat,
+            EnvVars.SystemAccessToken,
+        };
+
+        /// <summary>
+        /// Get either an ADO PAT from the environment. If one is not set, then do normal ADO AAD Auth.
+        /// </summary>
+        /// <param name="env">The <see cref="IEnv"/> to use for interacting with the environment.</param>
+        /// <returns>A PAT from the env if set, otherwise null.</returns>

```

Comment:
I think this is wrong. You never return `null`, it's always a `PatResult`. The `PatResult` just might have a `null` for `EnvVarSource` and `Value`.

Reply:
Goo catch, updated for the method this turned into :)